### DISCUSSION TITLE
Is *mail.Client goroutine safe?

### DISCUSSION THREAD

**Discussion by adpande** [wneessen/go-mail#268] [Q&A] [answered]
Can I create *mail.Client on top level once and use it in goroutines? Actually we are doing it and getting error:

```
panic: out of sync

goroutine 66897 [running]:
net/textproto.(*sequencer).End(0xc00095b020, 0xc000455698?)
/opt/hostedtoolcache/go/1.21.12/x64/src/net/textproto/pipeline.go:103 +0x145
net/textproto.(*Pipeline).EndResponse(0x182dc20?, 0xc00095af01?)
/opt/hostedtoolcache/go/1.21.12/x64/src/net/textproto/pipeline.go:65 +0x19
github.com/wneessen/go-mail/smtp.(*Client).cmd(0xc000d4e370, 0x0?, {0x1bc1e56, 0x4}, {0x0,
0x0, 0x0})
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/smtp/smtp.go:147 +0x279
github.com/wneessen/go-mail/smtp.(*Client).Quit(0xc000d4e370)
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/smtp/smtp.go:444 +0x45
github.com/wneessen/go-mail/smtp.(*Client).Auth(0xc000d4e370, {0x1e755c8, 0xc00099ef40})
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/smtp/smtp.go:210 +0x288
github.com/wneessen/go-mail.(*Client).auth(0xc0002ee800)
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/client.go:672 +0x665
github.com/wneessen/go-mail.(*Client).DialWithContext(0xc0002ee800, {0x1e80c10, 0x2f98e40}
)
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/client.go:532 +0x3b2
github.com/wneessen/go-mail.(*Client).DialAndSendWithContext(0xc0008e5420?, {0x1e80c10?, 0
x2f98e40?}, {0xc000455b30, 0x1, 0x1})
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/client.go:573 +0x35
github.com/wneessen/go-mail.(*Client).DialAndSend(...)
/home/runner/go/pkg/mod/github.com/wneessen/go-mail@v0.4.0/client.go:567
github.com/statstream/statstream/asgard/email.SendPlainTextMail({0xc000a0c150?, 0xc000455b
f0?}, {0xc00079e530, 0xc}, {0xc0001c8300, 0xba}, 0x0, {0xc000d861f8, 0x16})
/home/runner/work/asgard/asgard/email/email.go:161 +0x393
github.com/statstream/statstream/asgard/email.SendAlertUsingEmailPlainText({0xc000660e90,
0xd}, {0xc000a0c150, 0x28}, {0xc000d861f8, 0x16}, {0xc00079e530, 0xc}, {0xc00079e550, 0xc}
)
/home/runner/work/asgard/asgard/email/email.go:267 +0x210

```

**Comment by wneessen** [comment, marked as answer]
In this case it looks like `*smtp.Client` is the culprit. It makes use of maps which are not concurrency-safe. Since the `*mail.Client` uses the `*smtp.Client` it might fail. I've never really used go-mail in goroutines yet, so I would say it's not currency-safe at this point. I'll add an issue though, to look at this at some point.

**Reply by wneessen** [reply]
https://github.com/wneessen/go-mail/issues/269 has been filed

**Comment by adpande** [comment]
Thanks @wneessen for clarification. We are now creating a new client for each request. Not sure if this will have any performance implications.

**Reply by wneessen** [reply]
It shouldn't have any performance implications (except maybe GC) since creating a client will only create a struct in memory. The actual "work" happens when you dial/send through it.

**Comment by firefart** [comment]
Having the same issue after switching from `gopkg.in/gomail.v2` to this library. Is there a chance that you use `sync.Map` or a mutex inside the client to make it threadsafe @wneessen?

**Reply by wneessen** [reply]
It's definetely on the roadmap and the next thing i'll be working on, but given my current workload (at work :D ) I can't promise any ETA, yet.

**Reply by firefart** [reply]
Thanks! Having a quick look it seems the `Client` is currently not reusable at all because the `smtp.Client` also holds a connection in it's struct. Maybe this should be noted in the docs in the meantime. Thanks for looking into this!

**Comment by wneessen** [comment]
Hi everyone,

I've just pushed #307 into main, which implements goroutine-/thread-safety into go-mail. I would appreciate if some of you could give it a test run and provide any feedback.
